README.md
# Advanced Camera Views [![Steam](https://img.shields.io/steam/downloads/2021073822.svg?label=Steam&logo=steam)](https://steamcommunity.com/sharedfiles/filedetails/?id=2021073822)
 
This is the codebase for my Transport Fever 2 Mod "Advanced Camera Views"


https://steamcommunity.com/workshop/filedetails/?id=2021073822
https://www.transportfever.net/filebase/index.php?entry/5224-advanced-camera-views/


changelog.txt
2023-01-09: v1.9
- Fix Bug in configs/fakeBogies repair script, add log info

2022-11-27: v1.8
- Fix for mods with incomplete fakeBogies definition

2022-11-25: v1.7
- Fix Freezes caused by senseless metadata.railVehicle.configs definitions from some mods and the vanilla skoda tram (thanks to @EAT1963)

2022-11-17: v1.6
- enhanced checking of boundingBox definiton

2022-11-13: v1.5
- Fix Freezes for crew&passenger views, new option for inserting the views
thanks to @EAT1963 !

2020-03-30: v1.3 
- Bugfix

2020-03-11: v1.2 
- Bugfix

2020-03-11: v1.1 
- Release

2020-02-22: v0.1 
- Beta

res/scripts/vTube_tools.lua
------------------------------------------------------------------------------------
--	
------------------------------------------------------------------------------------
local function getValue(value, default)
	return (type(value) ~= type(default)) and default or value
end

------------------------------------------------------------------------------------
--	Prüft, ob "id" vom Typ "string" und nicht leer ("") ist.
------------------------------------------------------------------------------------
local function isValidId(id)
	return ((type(id) == "string") and (id ~= ""))
end

local vtTools = {
	getValue = getValue,
	isValidId = isValidId,
}

------------------------------------------------------------------------------------
--	Anlegen/Prüfen globaler Daten
------------------------------------------------------------------------------------
local function init()
	if (type(vacuumTubeGlobal) ~= "table") then
		vacuumTubeGlobal = {}
	end
	if (type(vacuumTubeGlobal.modParams) ~= "table") then
		vacuumTubeGlobal.modParams = {}
	end
end
init()

------------------------------------------------------------------------------------
--	Rückgabe globale Daten.
------------------------------------------------------------------------------------
vtTools.myGlobalData = function()
	return vacuumTubeGlobal
end

------------------------------------------------------------------------------------
--	Rückgabe der modParams.
--		Ist key leer ("") --> Rückgabe aller modParams; andernfalls Daten von key.
------------------------------------------------------------------------------------
vtTools.modParams = function(key)
	if isValidId(key) then
		if (type(vacuumTubeGlobal.modParams[key]) ~= "table") then
			vacuumTubeGlobal.modParams[key] = {}
		end
		return vacuumTubeGlobal.modParams[key]
	else
		return vacuumTubeGlobal.modParams
	end
end

------------------------------------------------------------------------------------
--	Speichert modParams des aktuellen Mods global ab.
--		Alle Werte werden auf Gültigkeit geprüft und um 1 erhöht.
--		Somit sind diese bereits fit für evtl. lua-Tabellenauswahl per Index. :)
------------------------------------------------------------------------------------
vtTools.saveModParams = function(modKey, modParams, defaulModParams)
	if (isValidId(modKey) and (type(modParams) == "table")) then
		local result = getValue(modParams[getCurrentModId()], {})
		for i, value in ipairs(getValue(defaulModParams, {})) do
			result[value.key] = getValue(result[value.key], getValue(value.defaultIndex, 0)) + 1
		end
		vacuumTubeGlobal.modParams[modKey] = result
	end
end

return vtTools

res/scripts/merk_modutil_1.lua
local merkVehicleUtil
local merkSettingsUtil
local version = 1
local minorVersion = 0

require "stringutil"

-- base functions
local function fileExists(fileName)
	local file = io.open(fileName, "r")
	if file then
		file:close()
		return true
	end
	return false
end

local function table_equal(t1, t2)
	if type(t1) ~= "table" or type(t2) ~= "table" then
		return false
	end
	local isEqual = true
	local checkedKeys = {}
	for k, v in pairs(t1) do
		if (t2[k] == nil) or (type(t2[k]) ~= type(v)) then
			isEqual = false
			break
		elseif type(v) == "table" then
			isEqual = table_equal(v, t2[k])
			if not isEqual then
				break
			end
		elseif v ~= t2[k] then
			isEqual = false
			break
		end
		checkedKeys[k] = true
	end
	if isEqual then
		for k, v in pairs(t2) do
			if not checkedKeys[k] then
				if (t1[k] == nil) or (type(t1[k]) ~= type(v)) then
					isEqual = false
					break
				elseif type(v) == "table" then
					isEqual = table_equal(v, t1[k])
					if not isEqual then
						break
					end
				elseif v ~= t1[k] then
					isEqual = false
					break
				end
			end
		end
	end
	return isEqual
end

-- Repaints
local function addMaterial(id, file, materials)
	if type(merkVehicleUtil) == "table" and type(merkVehicleUtil[id]) == "table" and type(materials) == "table" then
		if type(merkVehicleUtil[id].meshData) == "table" and merkVehicleUtil[id].meshData[file] then
			local mesh = merkVehicleUtil[id].meshData[file]
			local matConfig = {}
			if type(mesh.materials) == "table" then
				for i, subMesh in ipairs(mesh.materials) do
					local matConf = 0
					for k, material in ipairs(materials) do
						if type(subMesh) == "table" and type(material) == "table" and material[1] and material[2] then
							for j, mat in ipairs(subMesh) do
								if mat == material[1] and mat ~= material[2] then
									subMesh[#subMesh+1] = material[2]
									matConf = #subMesh-1
								end
								if mat == material[2] and (j < #subMesh or matConf == 0) then
									-- material already exists
									if matConf > 0 then
										subMesh[#subMesh] = nil
										matConf = j-1
									end
									break
								end
							end
						end
					end
					matConfig[#matConfig+1] = matConf
				end
			end
			
			local matConfigId = #mesh.matConfigs+1
			if type(mesh.matConfigs) == "table" and #matConfig > 0 then
				for i, matConf in ipairs(mesh.matConfigs) do
					if table_equal(matConf, matConfig) then
						matConfigId = i
						break
					end
				end
				if matConfigId > #mesh.matConfigs then
					mesh.matConfigs[matConfigId] = matConfig
				end
				return matConfigId-1
			end
		end
	end
	return nil
end

local function getGroupMatConfig(id, file, matConfigs)
	if type(merkVehicleUtil) == "table" and type(merkVehicleUtil[id]) == "table" then
		if type(merkVehicleUtil[id].groupData) == "table" and merkVehicleUtil[id].groupData[file] then
			local group = merkVehicleUtil[id].groupData[file]
			local matConfig = {}
			
			if type(group.meshes) == "table" and type(matConfigs) == "table" then
				for i, mesh in ipairs(group.meshes) do
					if matConfigs[mesh] then
						matConfig[#matConfig+1] = matConfigs[mesh]
					else
						matConfig[#matConfig+1] = 0
					end
				end
				
				if type(group.matConfigs) == "table" then
					local matConfigID = #group.matConfigs+1
					for i, matConf in ipairs(group.matConfigs) do
						if table_equal(matConf, matConfig) then
							matConfigID = i
							break
						end
					end
					if matConfigID > #group.matConfigs then
						group.matConfigs[matConfigID] = matConfig
					end
					return matConfigID-1
				end
			end
		end
	end
	return nil
end

local function addRepaint(id, repaintId, materials, files)
	if type(merkVehicleUtil) == "table" then
		if type(merkVehicleUtil[id]) == "table" and merkVehicleUtil[id].created then
			-- make sure "meshData" and "groupData" are tables
			local meshData = {}
			if type(merkVehicleUtil[id].meshData) == "table" then
				meshData = merkVehicleUtil[id].meshData
			end
			local groupData = {}
			if type(merkVehicleUtil[id].groupData) == "table" then
				groupData = merkVehicleUtil[id].groupData
			end
			
			local meshFiles = {}
			local groupFiles = {}
			if type(files) == "table" and #files > 0 then
				for i, file in ipairs(files) do
					if meshData[file] then
						meshFiles[#meshFiles+1] = file
					elseif groupData[file] then
						groupFiles[#groupFiles+1] = file
					end
				end
			else
				for file, data in pairs(meshData) do
					meshFiles[#meshFiles+1] = file
				end
				for file, data in pairs(groupData) do
					groupFiles[#groupFiles+1] = file
				end
			end
			
			local matConfigs = {}
			for i, file in ipairs(meshFiles) do
				matConfigs[file] = addMaterial(id, file, materials)
			end
			
			for i, file in ipairs(groupFiles) do
				matConfigs[file] = getGroupMatConfig(id, file, matConfigs)
			end
			
			if type(merkVehicleUtil[id].repaints) ~= "table" then
				merkVehicleUtil[id].repaints = {}
			end
			
			merkVehicleUtil[id].repaints[repaintId] = matConfigs
		else
			-- main mod not loaded, store data for later use
			if type(merkVehicleUtil[id]) ~= "table" then
				merkVehicleUtil[id] = {}
			end
			if type(merkVehicleUtil[id].pendingRepaints) ~= "table" then
				merkVehicleUtil[id].pendingRepaints = {}
			end
			merkVehicleUtil[id].pendingRepaints[#merkVehicleUtil[id].pendingRepaints+1] = {
				id = repaintId,
				materials = materials,
				files = files
			}
		end
	end
end

local function getRepaint(id, repaintId, data)
	if type(merkVehicleUtil) == "table" and type(merkVehicleUtil[id]) == "table" and type(merkVehicleUtil[id].repaints) == "table" then
		if type(merkVehicleUtil[id].repaints[repaintId]) == "table" then
			local matConfigs = merkVehicleUtil[id].repaints[repaintId]
		
			if type(data) == "table" and type(data.lods) == "table" then
				for i, lod in ipairs(data.lods) do
					local matConfig = {}
					if type(lod) == "table" then
						for j, child in ipairs(lod.children) do
							if matConfigs[child.id] then
								matConfig[#matConfig+1] = matConfigs[child.id]
							elseif lod.matConfigs[1][j] then
								matConfig[#matConfig+1] = lod.matConfigs[1][j]
							else
								matConfig[#matConfig+1] = 0
							end
						end
					else
						print("[ModUtil] Missing key \"lod.children\".")
					end
					lod.matConfigs[1] = matConfig
				end
			end
	
			return matConfigs
		end
	end
	if type(data) == "table" then
		data.metadata = {}
	end
	
	return {}
end

local function getMeshData(id, fileName, data)
	if type(merkVehicleUtil) == "table" and type(merkVehicleUtil[id]) == "table" and type(merkVehicleUtil[id].meshData) == "table" then
		local meshData = merkVehicleUtil[id].meshData[fileName]
		
		if type(meshData) == "table" and data then
			if meshData.matConfigs then
				data.matConfigs = meshData.matConfigs
			end
			if type(meshData.materials) == "table" and type(data.subMeshes) == "table" then
				for i, subMesh in ipairs(data.subMeshes) do
					if meshData.materials[i] then
						subMesh.materials = meshData.materials[i]
					end
				end
			end
		end
		
		return meshData
	end
end

local function getGroupData(id, fileName, data)
	if type(merkVehicleUtil) == "table" and type(merkVehicleUtil[id]) == "table" and type(merkVehicleUtil[id].groupData) == "table" then
		local groupData = merkVehicleUtil[id].groupData[fileName]
		
		if type(groupData) == "table" and data then
			if groupData.matConfigs then
				data.matConfigs = groupData.matConfigs
			end
		end
		
		return groupData
	end
end

local function setMetadata(id, fileName, metadata, priority)
	if type(merkVehicleUtil) == "table" then
		if type(merkVehicleUtil[id]) ~= "table" then
			merkVehicleUtil[id] = {}
		end		
		
		local prio = priority or 0
		if type(merkVehicleUtil[id].metadata) ~= "table" then
			merkVehicleUtil[id].metadata = {}
		end
		if type(merkVehicleUtil[id].metadata[fileName]) ~= "table" then
			merkVehicleUtil[id].metadata[fileName] = {}
		end
		
		metadataCache = merkVehicleUtil[id].metadata[fileName]
		
		if type(metadata) == "table" then
			for key, value in pairs(metadata) do
				if metadataCache[key] == nil or (type(metadataCache[key]) == "table" and metadataCache[key].priority < prio) then
					metadataCache[key] = {
						value = value,
						priority = prio
					}
				end
			end
		end
	end
end

local function changeMetadata(id, fileName, data)
	if type(merkVehicleUtil) == "table" and type(merkVehicleUtil[id]) == "table" then
		local metadata = {}
		if type(merkVehicleUtil[id].metadata) == "table" and type(merkVehicleUtil[id].metadata[fileName]) == "table" then
			metadata = merkVehicleUtil[id].metadata[fileName]
		end
		
		if type(data) == "table" and type(data.metadata) == "table" then
			for key, value in pairs(metadata) do
				data.metadata[key] = value.value
			end
		end
	end
end

local function readFiles(files, options)
	local meshData = {}
	local groupData = {}
	local dataBackup = data
	
	local info
	local i = 1
	repeat
		info = debug.getinfo(i,'S')
		i = i+1
	until info == nil
	info = debug.getinfo(i-2,'S')
	
	local modPath
	if string.ends(info.source, 'mod.lua') then
		modPath = string.gsub(info.source, "@(.*/)mod[.]lua", "%1")
	elseif string.ends(info.source, '.mdl') then
		modPath = string.gsub(info.source, "@(.*/)res/models/model/.+[.]mdl", "%1")
	end
	
	if fileExists(modPath.."mod.lua") then
		local repaintFileName = modPath.."repaintFiles.lua"
		if type(options) ~= "table" or (options.useRepaintFile ~= false and options.devMode ~= true) then
			local repaintFile = loadfile(repaintFileName)
			if repaintFile then
				return repaintFile()
			end
		end
		
		if type(options) == "table" and (options.loadMeshes or (options.loadMeshes == nil and options.devMode)) then
			for i, file in ipairs(files) do
				if type(file) == "string" then
					if string.ends(file, ".msh") then
						local meshFile = loadfile(modPath.."res/models/mesh/"..file)
						if meshFile then
							meshFile()
						end
						if meshFile and type(data) == "function" then
							local mesh = data()
							local mData
							
							if type(mesh) == "table" and mesh.matConfigs and mesh.subMeshes then
								mData = {
									materials = {},
								}
								mData.matConfigs = mesh.matConfigs
								for j, subMesh in ipairs(mesh.subMeshes) do
									if subMesh.materials then
										mData.materials[#mData.materials+1] = subMesh.materials
									else
										print(string.format("[VehicleUtil] Missing field \"submesh.material\" in mesh %q!"), file)
									end
								end
							end
							
							meshData[file] = mData
						end
					elseif string.ends(file, ".grp") then
						local groupFile = loadfile(modPath.."res/models/group/"..file)
						if groupFile then
							groupFile()
						end
						if groupFile and type(data) == "function" then
							local group = data()
							local gData
							
							if type(group) == "table" and group.matConfigs and group.children then
								gData = {
									meshes = {},
								}
								gData.matConfigs = group.matConfigs
								for j, child in ipairs(group.children) do
									if child.id then
										gData.meshes[#gData.meshes+1] = child.id
									else
										print(string.format("[VehicleUtil] Missing field \"child.id\" in group %q!"), file)
									end
								end
							end
							
							groupData[file] = gData
						end
					end
				end
			end
			
			data = dataBackup
			
			if options.createRepaintFile or (options.createRepaintFile == nil and options.devMode) then
				local repaintFile = io.open(repaintFileName, "w")
				if repaintFile then
					require "serialize"
					repaintFile:write("local "..serializeStr({ meshData = meshData, groupData = groupData }))
					repaintFile:write("local result = data()\nreturn result.meshData, result.groupData")
					repaintFile:close()
				else
					print("[ModUtil] Creation of repaint file failed")
				end
			end
		end
	else
		print("[ModUtil] Could not read files. Maybe the function was called from a wrong file.")
	end
	
	return meshData, groupData
end

local function createEnvironment(id, files, meshes, groups, options, modelViewer)
	local meshData = {}
	local groupData = {}
	if type(files) == "table" and #files > 0 then
		meshData, groupData = readFiles(files, options)
	else
		meshData = meshes
		groupData = groups
	end
	
	if game or modelViewer then
		if modelViewer then
			if merkVehicleUtil == nil then
				merkVehicleUtil = {}
			end
		else
			if type(game.merkVehicleUtil) ~= "table" then
				game.merkVehicleUtil = {}
			end
			if type(game.merkVehicleUtil[version]) ~= "table" then
				game.merkVehicleUtil[version] = {}
			end
			merkVehicleUtil = game.merkVehicleUtil[version]
		end
		
		if type(merkVehicleUtil[id]) == "table" and merkVehicleUtil[id].created then
			print(string.format("[ModUtil] Creation of repaint environment for id %q failed, the id is already registered.", id))
		else
			if type(merkVehicleUtil[id]) ~= "table" then
				merkVehicleUtil[id] = {}
			end
			
			merkVehicleUtil[id].created = true
			
			merkVehicleUtil[id].meshData = meshData
			merkVehicleUtil[id].groupData = groupData
			
			if type(merkVehicleUtil[id].pendingRepaints) == "table" and #merkVehicleUtil[id].pendingRepaints > 0 then
				pendingRepaints = merkVehicleUtil[id].pendingRepaints
				merkVehicleUtil[id].pendingRepaints = nil
				for i, repaint in ipairs(pendingRepaints) do
					addRepaint(id, repaint.id, repaint.materials, repaint.files)
				end
			end
			
			return {
				addRepaint = function(repaintId, materials, files)
					return addRepaint(id, repaintId, materials, files)
				end,
				getRepaint = function(repaintId, data)
					return getRepaint(id, repaintId, data)
				end,
				getMeshData = function(fileName)
					return getMeshData(id, fileName)
				end,
				getGroupData = function(fileName)
					return getGroupData(id, fileName)
				end,
				setMetadata = function(fileName, metadata, priority)
					return setMetadata(id, fileName, metadata, priority)
				end,
				changeMetadata = function(fileName, data)
					return changeMetadata(id, fileName, data)
				end,
			}
		end
	else
		print(string.format("[ModUtil] Creation of repaint environment for id %q failed. It seems you're trying to create the environment outside the \"runFn\".", id))
	end
end

local function modelViewerRepaint(id, repaintId, materials, files, data)
	createEnvironment(id, {""}, nil, nil, nil, true, 4)
	addRepaint(id, repaintId, materials, files)
	getRepaint(id, repaintId, data)
end

if not merk_vehicleUtil or not merk_vehicleUtil[version] then
	if game then
		if type(game.merkVehicleUtil) ~= "table" then
			game.merkVehicleUtil = {}
		end
		if type(game.merkVehicleUtil[version]) ~= "table" then
			game.merkVehicleUtil[version] = {}
		end
		merkVehicleUtil = game.merkVehicleUtil[version]
	end

	if type(merk_vehicleUtil) ~= "table" then
		merk_vehicleUtil = {}
	end

	merk_vehicleUtil[version] = {
		createEnvironment = createEnvironment,
		addRepaint = addRepaint,
		getRepaint = getRepaint,
		getMeshData = getMeshData,
		getGroupData = getGroupData,
		modelViewerRepaint = modelViewerRepaint,
		setMetadata = setMetadata,
		changeMetadata = changeMetadata,
		minorVersion = minorVersion,
	}
else
	if merk_vehicleUtil[version].minorVersion < minorVersion then
		merk_vehicleUtil[version] = {
			createEnvironment = createEnvironment,
			addRepaint = addRepaint,
			getRepaint = getRepaint,
			getMeshData = getMeshData,
			getGroupData = getGroupData,
			modelViewerRepaint = modelViewerRepaint,
			setMetadata = setMetadata,
			changeMetadata = changeMetadata,
			minorVersion = minorVersion,
		}
	end
end

-- settings
local function readUserSettings(id)	
	if type(merkSettingsUtil) == "table" and type(merkSettingsUtil[id]) == "table" then
		local settings = merkSettingsUtil[id].values
		
		local info
		local i = 1
		repeat
			info = debug.getinfo(i,'S')
			i = i+1
		until info == nil
		info = debug.getinfo(i-2,'S')
		
		local modPath
		if string.ends(info.source, 'mod.lua') then
			modPath = string.gsub(info.source, "@(.*/)mod[.]lua", "%1")
		elseif string.ends(info.source, '.mdl') then
			modPath = string.gsub(info.source, "@(.*/)res/models/model/.+[.]mdl", "%1")
		end
		
		if fileExists(modPath.."mod.lua") then
			local settingsFile = loadfile(modPath.."settings.lua")
			local loadedSettings = {}
			
			if settingsFile then
				loadedSettings = settingsFile()
			end
			
			if type(loadedSettings) == "table" then
				for option_key, option in pairs(merkSettingsUtil[id].definitions) do
					if type(loadedSettings[option_key]) == option.type then
						local value = loadedSettings[option_key]
						if option.type == "number" then
							if option.min and value < option.min then
								value = option.min
							end
							if option.max and value > option.max then
								value = option.max
							end
						end
						settings[option_key] = value
					end
				end
			end
		end
	end
end

local function createUserSettings(id, options)
	
	if type(merkSettingsUtil) == "table" then
		if merkSettingsUtil[id] then
			print(string.format("[ModUtil] Previous settings for id %q are overwritten.", id))
		end
		
		merkSettingsUtil[id] = {
			definitions = options,
			values = {},
		}
		
		for option_key, option in pairs(options) do
			if option.type and option.name then
				if option.default ~= nil then
					merkSettingsUtil[id].values[option_key] = option.default
				else
					if option.type == "boolean" then
						merkSettingsUtil[id].values[option_key] = false
					elseif option.type == "number" then
						merkSettingsUtil[id].values[option_key] = 0
					elseif option.type == "string" then
						merkSettingsUtil[id].values[option_key] = ""
					end
				end
			end
		end
		
		readUserSettings(id)
	end
end

local function getUserSettings(id)
	local settings = {}
	
	if type(merkSettingsUtil) == "table" and type(merkSettingsUtil[id]) == "table" then
		settings = merkSettingsUtil[id].values
	end
	
	return settings
end

if not merk_SettingsUtil or not merk_SettingsUtil[version] then
	if game then
		if type(game.merkSettingsUtil) ~= "table" then
			game.merkSettingsUtil = {}
		end
		if type(game.merkSettingsUtil[version]) ~= "table" then
			game.merkSettingsUtil[version] = {}
		end
		merkSettingsUtil = game.merkSettingsUtil[version]
	end

	if type(merk_SettingsUtil) ~= "table" then
		merk_SettingsUtil = {}
	end

	merk_SettingsUtil[version] = {
		create = createUserSettings,
		get = getUserSettings,
		minorVersion = minorVersion,
	}
else
	if merk_SettingsUtil[version].minorVersion < minorVersion then
		merk_SettingsUtil[version] = {
			create = createUserSettings,
			get = getUserSettings,
			minorVersion = minorVersion,
		}
	end
end

local mod_util = {
	vehicles = merk_vehicleUtil[version],
	userSettings = merk_SettingsUtil[version],
	initialize = function (id, settings, files, meshes, groups, options, modelViewer)
			merk_SettingsUtil[version].create(id, settings)
			return merk_vehicleUtil[version].createEnvironment(id, files, meshes, groups, options, modelViewer)
		end,
	minorVersion = minorVersion,
}

if not merk_modutil or not merk_modutil[version] then
	if type(merk_modutil) ~= "table" then
		merk_modutil = {}
	end
	merk_modutil[version] = mod_util
else
	if merk_modutil[version].minorVersion < minorVersion then
		merk_modutil[version] = mod_util
	end
end

return merk_modutil[version]


res/scripts/addCameraPositions.lua
local tools = require "vTube_tools"
local camViews = require "AdvancedCameraViews"

local function addPositions (data, views)  -- add Camera view(s) to existing ones
	local function doInsert(source, dest)
		if source.transf then
			table.insert(dest, source)
		elseif (type(source) == "table") then
			for i, value in ipairs(source) do
				doInsert(value, dest)
			end
		end
	end
	----------------------------
	local options = tools.modParams("paramsACV")
	if ((options.replaceExistingViews == 1) or (type(data.metadata.cameraConfig) ~= "table") or (type(data.metadata.cameraConfig.positions) ~= "table")) then
		-- options.replaceExistingViews == 1	--> Vorhandene ersetzen
		data.metadata.cameraConfig = {positions = {}}
	end
	
	if (options.replaceExistingViews == 2) then
		-- options.replaceExistingViews == 2	--> Neue am Anfang platzieren
		local positions = {}
		doInsert({views, data.metadata.cameraConfig.positions}, positions)
		data.metadata.cameraConfig.positions = positions
	else
		-- options.replaceExistingViews == 3	--> Neue am Ende platzieren
		doInsert(views, data.metadata.cameraConfig.positions)
	end
end

local deactivatableKeys = {
	"backBackwardParts",
	"backForwardParts",
	"frontBackwardParts",
	"frontForwardParts",
	"innerBackwardParts",
	"innerForwardParts",
	"blinkingLights0",
	"blinkingLights1",
	"blinkLightsRight0",
	"blinkLightsRight1",
	"blinkLightsLeft0",
	"blinkLightsLeft1",
	"brakeLights",
}

local function checkFakeBogie(bogie, countAxles)
	local group = tools.getValue(bogie.group, -1)
	local sum = tools.getValue(bogie.offset, 0) ~= 0 or tools.getValue(bogie.position, 0) ~= 0
	return group > 0 or (group == 0 and (sum == true or (sum == false and countAxles == 0)))
end

-- fix some senseless definitions from some mods leading to errors
local function repairConfigs(configs, fileName)
	for i = 1, #configs do
		--	handling deactivatableKeys (no zero entrys)
		for j, deactivatableKey in ipairs(deactivatableKeys) do
			local newKeys = {}
			for k, id in ipairs(tools.getValue(configs[i][deactivatableKey], {})) do
				if (id > 0) then
					table.insert(newKeys, id)
				else
					print(string.format("Advanced Camera Views - Delete 0 entry in 'configs[%d].%s' [%s].", i, deactivatableKey, fileName))
				end
			end
			configs[i][deactivatableKey] = newKeys
		end
		--	handling fakeBogies
		local newFakeBogies = {}
		for j, bogie in ipairs(tools.getValue(configs[i].fakeBogies, {})) do
			if checkFakeBogie(bogie, #tools.getValue(configs[i].axles, {})) then
				table.insert(newFakeBogies, bogie)
			else
				print(string.format("Advanced Camera Views - Delete invalid fakeBogie: 'configs[%d].fakeBogies[%d].' [%s]", i, j, fileName))
			end
		end
		configs[i].fakeBogies = newFakeBogies
	end
end

local function createGroupsTable(fileName, data)
	local function handleGroup(result, params)
		local copyKeys = {
			"materials",
			"mesh",
			"name",
			"transf",
		}
		params.currentId = params.currentId + 1
		local myId = params.currentId
		local entry = {
			id = params.currentId,
			parentId = params.parentId,
			data = {},
		}
		for i, key in ipairs(copyKeys) do
			entry.data[key] = params.root[key]
		end
		result.childrenData[myId] = entry
		
		if (type(params.root.children) == "table") then
			for i, child in ipairs(params.root.children) do
				params.currentId = handleGroup(result, {
						root = child,
						currentId = params.currentId,
						parentId = myId,
					} )
			end
		end
		
		return params.currentId
	end
	--------------------------------------------------
	-- Hier geht's los :)
	--------------------------------------------------
	local result = {
		childrenData = {},
		deactivatableIds = {},
	}
	
	if ((type(data.lods) == "table") and (type(data.lods[1]) == "table") and (type(data.lods[1].node) == "table")) then
		--	Group- and mesh handling
		handleGroup(result, {
				root = data.lods[1].node,
				currentId = -1,
				parentId = -1,
			} )
			
		--	Handling deactivatable groups
		if ((type(data.metadata.railVehicle) == "table") and (type(data.metadata.railVehicle.configs) == "table") and
			(type(data.metadata.railVehicle.configs[1]) == "table")) then
			
			repairConfigs(data.metadata.railVehicle.configs, fileName)
			
			for i, deactivatableKey in ipairs(deactivatableKeys) do
				for j, id in ipairs(tools.getValue(data.metadata.railVehicle.configs[1][deactivatableKey], {})) do
					result.deactivatableIds[id] = true
				end
			end
		end
	end
	
	return result
end

local function isValidBoundingInfo(boundingInfo)
	-- Sicherstellen, dass "boundingInfo" valid ist und das Modell 3-dimensionale Ausmaße hat.
	if ((type(boundingInfo) == "table") and (type(boundingInfo.bbMax) == "table") and (type(boundingInfo.bbMin) == "table")) then
		if ((boundingInfo.bbMax[1] - boundingInfo.bbMin[1] > 0) and (boundingInfo.bbMax[2] - boundingInfo.bbMin[2] > 0) and
			(boundingInfo.bbMax[3] - boundingInfo.bbMin[3] > 0)) then
			return true
		end
	end
	return false
end


local invalidModels = {
	-- ["res/models/model/vehicle/tram/usa/skoda_10t.mdl"] = true,  -- solved with repairConfigs
}

-- addModifier "loadModel"
return function (fileName, data)
	-- if invalidModels[fileName] then
		-- return data
	-- end
	
	if data and data.metadata and isValidBoundingInfo(data.boundingInfo) then
		local options = tools.modParams("paramsACV")
		local viewsParams = {
			fileName = fileName,
			data = data,
			groupsTable = createGroupsTable(fileName, data),
		}
		
		local xmax=data.boundingInfo.bbMax[1]
		local ymax=data.boundingInfo.bbMax[2]
		local zmax=data.boundingInfo.bbMax[3]
		local xmin=data.boundingInfo.bbMin[1]
		local ymin=data.boundingInfo.bbMin[2]
		local zmin=data.boundingInfo.bbMin[3]
		
		if data.metadata.transportVehicle then
			if data.metadata.transportVehicle.carrier=="RAIL" and data.metadata.railVehicle then  -- Rail vehicle
				if data.metadata.railVehicle.engines and #data.metadata.railVehicle.engines>0 then
					addPositions(data, camViews.TrainViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams)) -- Traction unit
				else
					addPositions(data, camViews.WaggonViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams))  -- Waggon
				end
			end
			if data.metadata.transportVehicle.carrier=="TRAM" and data.metadata.railVehicle then  -- Tram vehicle
				if data.metadata.railVehicle.engines then
					addPositions(data, camViews.TramViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams))
				else
					
				end
			end
			if data.metadata.transportVehicle.carrier=="ROAD" and data.metadata.roadVehicle then  -- Road vehicle
				addPositions(data, camViews.RoadViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams))
			end
			if data.metadata.waterVehicle then  -- Ship
				addPositions(data, camViews.ShipViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams))
			end
			if data.metadata.airVehicle then  -- Airplane
				addPositions(data, camViews.PlaneViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams))
			end
		end
		if data.metadata.car then  -- Car
			addPositions(data, camViews.CarViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams))
		end
		if (data.metadata.person and (options.addPersonViews == 1)) then  -- Person
			addPositions(data, camViews.PersonViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams))
		end
		if data.metadata.animal then  -- Animal
			addPositions(data, camViews.AnimalViews(xmax,ymax,zmax,xmin,ymin,zmin,viewsParams))
		end
		-- data.metadata.cameraConfig = {positions = { {} }} -- try this for fun
	end
	return data
end

mod.lua
local tools = require "vTube_tools"

local crewPositions = {0}
local function setCrewPositions(entry)
	entry.values = {tostring(0)}
	local step = 0.05
	local count = 30
	local default = -0.8
	for i = 1, count do
		table.insert(crewPositions, step * i)
		table.insert(entry.values, tostring(step * i))
		table.insert(crewPositions, 1, -1 * step * i)
		table.insert(entry.values, 1, tostring(-1 * step * i))
	end
	entry.defaultIndex = (default / step) + count
end

local defaultModParams = (function()
	local result = {}
	result[#result + 1] = {
		key = "replaceExistingViews",
		name = _("replaceExistingViewsNAME"),
		tooltip = _("replaceExistingViewsTT"),
		uiType = "COMBOBOX",
		values = { _("Replace existing"), _("Place new at the beginning"), _("Place new at the end")},
		defaultIndex = 1,
	}
	result[#result + 1] = {
		key = "addCrewSeats",
		name = _("addCrewSeatsNAME"),
		tooltip = _("addCrewSeatsTT"),
		uiType = "BUTTON",
		values = { _("Yes"), _("No")},
		defaultIndex = 0,
	}
	-- result[#result + 1] = {
		-- key = "posCrewSeats",
		-- name = _("posCrewSeatsNAME"),
		-- tooltip = _("posCrewSeatsTT"),
		-- uiType = "SLIDER",
	-- }
	-- setCrewPositions(result[#result])
	result[#result + 1] = {
		key = "addPassengerSeats",
		name = _("addPassengerSeatsNAME"),
		tooltip = _("addPassengerSeatsTT"),
		uiType = "BUTTON",
		values = { _("Yes"), _("No")},
		defaultIndex = 0,
	}
	result[#result + 1] = {
		key = "addPersonViews",
		name = _("addPersonViewsNAME"),
		tooltip = _("addPersonViewsTT"),
		uiType = "BUTTON",
		values = { _("Yes"), _("No")},
		defaultIndex = 1,
	}
	return result
end)()

function data()
	return {
		info = {
			name = _("name"),
			description = _("desc"),
			minorVersion = 9,
			severityAdd = "NONE",
			severityRemove = "NONE",
			tfnetId = 5224,
			tags = {"Script Mod","Camera","Camera View","Vehicle","Locomotive","Multiple Unit","Bus","Truck","Tram","Ship","Plane","Car","Person","Animal","Misc"},
			authors = {
				{
					name = "VacuumTube",
					role = "CREATOR",
					tfnetId = 29264,
				},                             
				{
					name = "EAT1963",
					role = "CO_CREATOR",
					tfnetId = 19725,
				},
			},
			params = defaultModParams,
		},
		runFn = function(settings, modParams)
			tools.saveModParams("paramsACV", modParams, defaultModParams)
			
			local entry = {}
			setCrewPositions(entry)
			tools.modParams("paramsACV").posCrewSeats = entry.defaultIndex
			tools.modParams("paramsACV").crewPositions = crewPositions
			
			addModifier("loadModel", require "addCameraPositions")
		end
	}
end